api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+					break
+				}
+			}
+
+			if !hasRoom {
+				newProvider, err := e.calculateNewProvider(clusterName)
+				if err != nil {
+					s.Errs = append(s.Errs, err)
+					continue
+				}
+
+				if !newProvider.HasResourcesFor(consumer) {
+					text := fmt.Sprintf("Resource '%s' cannot fit into an empty provider!", consumer.ID)
+					text += "\nThe instance size in your environment is too small to run this resource."
+					text += "\n Please increase the instance size for your environment."
+					err := fmt.Errorf(text)
+					s.Errs = append(s.Errs, err)
+					continue
+				}
+
+				newProvider.SubtractResourcesFor(consumer)
+				s.Providers = append(s.Providers, newProvider)
+			}
+		}
+	}
+
+	if len(sorts[0].Providers) > len(sorts[1].Providers) {
+		return sorts[0].Providers, sorts[0].Errs
+	}
+
+	return sorts[1].Providers, sorts[1].Errs
+}
+
+func (e *EnvironmentScaler) getContainerResourceFromDeploy(deployID string) ([]ResourceConsumer, error) {
+	// use some kind of deploy cache
+	if consumers, ok := e.deployCache[deployID]; ok {
+		return consumers, nil
+	}
+
+	input := &ecs.DescribeTaskDefinitionInput{}
+	input.SetTaskDefinition(deployID)
+	output, err := e.Client.ECS.DescribeTaskDefinition(input)
+	if err != nil {
+		return nil, err
+	}
+
+	consumers := make([]ResourceConsumer, len(output.TaskDefinition.ContainerDefinitions))
+	for i, d := range output.TaskDefinition.ContainerDefinitions {
+		var memory bytesize.Bytesize
+
+		if aws.Int64Value(d.MemoryReservation) != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(aws.Int64Value(d.MemoryReservation))
+		}
+
+		if aws.Int64Value(d.Memory) != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(aws.Int64Value(d.Memory))
+		}
+
+		ports := []int{}
+		for _, p := range d.PortMappings {
+			if aws.Int64Value(p.HostPort) != 0 {
+				ports = append(ports, int(aws.Int64Value(p.HostPort)))
+			}
+		}
+
+		consumers[i] = ResourceConsumer{
+			ID:     "",
+			Memory: memory,
+			Ports:  ports,
+		}
+	}
+
+	e.deployCache[deployID] = consumers
+	return consumers, nil
+}
+
+func (e *EnvironmentScaler) getResourceConsumers_PendingServices(clusterName string) ([]ResourceConsumer, error) {
+	var resourceConsumers []ResourceConsumer
+
+	listServicesInput := &ecs.ListServicesInput{}
+	listServicesInput.SetCluster(clusterName)
+
+	serviceARNs := []*string{}
+	listServicesPagesFN := func(output *ecs.ListServicesOutput, lastPage bool) bool {
+		serviceARNs = append(serviceARNs, output.ServiceArns...)
+
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListServicesPages(listServicesInput, listServicesPagesFN); err != nil {
+		return nil, err
+	}
+
+	services := []*ecs.Service{}
+	if len(serviceARNs) > 0 {
+		// The SDK states that you can specify up to 10 services in one DescribeServices operation:
+		// https://github.com/aws/aws-sdk-go/blob/ee1f179877b2daf2aaabf71fa900773bf8842253/service/ecs/api.go#L5420
+		// (aws-sdk-go version 1.12.19, as stated in layer0/Gopkg.toml)
```

Comment:
[nit] linked file is v1.10.51

Reply:
It's actually confusingly from the v1.12.19 tag. However, 'v1.10.51' is, I guess, the last commit message for this file. For less confusing, I'll update the link to be `.../blob/v.1.12.19/...` instead of `.../blob/LongCommitHash/...`.

✅